fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordContext.java

```diff
@@ -235,6 +259,110 @@ public Transaction ensureActive() {
         return transaction;
     }
 
+    /**
+     * Set the read version used by this transaction. All reads to the database will include
+     * only changes that were committed at this version or smaller. A transaction's read version
+     * can only be set once, so if this function is called multiple times, it will return the
+     * previously set read version. If this method is called and another caller has already called
+     * {@link #getReadVersionAsync()} or {@link #getReadVersion()}, this method may throw a
+     * {@link RecordCoreException} indicating that there is already an outstanding read version request
+     * if that request has not yet completed.
+     *
+     * @param readVersion the read version this transaction should use if is not already set
+     * @return this transaction's read version
+     * @see Transaction#setReadVersion(long)
+     */
+    public synchronized long setReadVersion(long readVersion) {
+        if (hasReadVersion()) {
+            return this.readVersion;
+        }
+        if (readVersionFuture != null) {
+            if (MoreAsyncUtil.isCompletedNormally(readVersionFuture)) {
+                return joinNow(readVersionFuture);
+            } else {
+                throw new RecordCoreException("Cannot set read version as read version request is outstanding");
+            }
+        }
+        ensureActive().setReadVersion(readVersion);
+        this.readVersion = readVersion;
+        this.readVersionFuture = CompletableFuture.completedFuture(readVersion);
+        return readVersion;
+    }
+
+    /**
+     * Get the read version used by this transaction. All reads to the database will include only changes that
+     * were committed at this version or smaller. If the read version has not already been set, this
+     * may require talking to the database. If the read version has already been set, then this will return
+     * with an already completed future.
+     *
+     * <p>
+     * Note that this method is {@code synchronized}, but only creating the future (<em>not</em> waiting on
+     * the future) will block other threads. Thus, while it is advised that this method ony be called once
+     * and by only one caller at a time, if it safe to use this method in asynchronous contexts.
+     * </p>
```

Comment:
I'm not sure that it is clear here that only one future will be created. Although, I don't know how to make it clearer, other than explicitly adding:
```suggestion
     * Note that a single future will be created, even if this is called many times.</p>
```